src/app/fdctl/run/tiles/fd_verify.c

```diff
@@ -0,0 +1,197 @@
+}
+
+FD_FN_CONST static inline void *
+mux_ctx( void * scratch ) {
+  return (void*)fd_ulong_align_up( (ulong)scratch, alignof( fd_verify_ctx_t ) );
+}
+
+/* during_frag is called between pairs for sequence number checks, as
+   we are reading incoming frags.  We don't actually need to copy the
+   fragment here, see fd_dedup.c for why we do this.*/
+
+static inline void
+during_frag( void * _ctx,
+             ulong in_idx,
+             ulong sig,
+             ulong chunk,
+             ulong sz,
+             int * opt_filter ) {
+  fd_verify_ctx_t * ctx = (fd_verify_ctx_t *)_ctx;
+
+  /* This is a dummy mcache entry to keep frags from getting overrun, do
+     not process */
+  if( FD_UNLIKELY( sig ) ) {
+    *opt_filter = 1;
+    return;
+  }
+
+  if( FD_UNLIKELY( chunk<ctx->in[in_idx].chunk0 || chunk>ctx->in[in_idx].wmark || sz > FD_TPU_DCACHE_MTU ) )
+    FD_LOG_ERR(( "chunk %lu %lu corrupt, not in range [%lu,%lu]", chunk, sz, ctx->in[in_idx].chunk0, ctx->in[in_idx].wmark ));
+
+  uchar * src = (uchar *)fd_chunk_to_laddr( ctx->in[in_idx].mem, chunk );
+  uchar * dst = (uchar *)fd_chunk_to_laddr( ctx->out_mem, ctx->out_chunk );
+
+  fd_memcpy( dst, src, sz );
+}
+
+static inline void
+after_frag( void *             _ctx,
+            ulong *            opt_sig,
+            ulong *            opt_chunk,
+            ulong *            opt_sz,
+            int *              opt_filter,
+            fd_mux_context_t * mux ) {
+  (void)opt_sig;
+  (void)mux;
+
+  fd_verify_ctx_t * ctx = (fd_verify_ctx_t *)_ctx;
+
+  uchar * udp_payload = (uchar *)fd_chunk_to_laddr( ctx->out_mem, ctx->out_chunk );
+  ushort payload_sz = *(ushort*)(udp_payload + *opt_sz - sizeof(ushort));
+  fd_txn_t * txn = (fd_txn_t*) fd_ulong_align_up( (ulong)(udp_payload) + payload_sz, 2UL );
+
+  ulong const * public_key = (ulong const *)(udp_payload + txn->acct_addr_off);
+  ulong const * sig        = (ulong const *)(udp_payload + txn->signature_off);
+  uchar const * msg        = (uchar const *)(udp_payload + txn->message_off);
+  ulong msg_sz             = (ulong)payload_sz - txn->message_off;
+
+  /* Sig is already effectively a cryptographically secure hash of
+     public_key/private_key and message and sz.  So use this to do a
+     quick dedup of ha traffic (FIXME: POTENTIAL DOS ATTACK IF
+     SOMEBODY COULD INTERCEPT TRAFFIC AND SUBMIT PACKETS WITH SAME
+     PUBLIC KEY, SIG AND GARBAGE MESSAGE AHEAD OF THE TRAFFIC ...
+     SEEMS UNLKELY AS THEY WOULD EITHER BE BEHIND THE INBOUND OR BE
+     A MITM THAT COULD JUST DISCARD INBOUND TRAFFIC). */
+
+  int ha_dup;
+  FD_TCACHE_INSERT( ha_dup, *ctx->tcache_sync, ctx->tcache_ring, ctx->tcache_depth, ctx->tcache_map, ctx->tcache_map_cnt, *sig );
+  if( FD_UNLIKELY( ha_dup ) ) {
+    *opt_filter = 1;
+    return;
+  }
+
+  /* We appear to have a message to verify.  So verify it. */
+
+  *opt_filter = !!fd_ed25519_verify( msg, msg_sz, sig, public_key, ctx->sha );
```

Comment:
```suggestion
  *opt_filter = (FD_ED25519_SUCCESS != fd_ed25519_verify( msg, msg_sz, sig, public_key, ctx->sha ));
```
I think this is equivalent and more explicit, but please double check.